mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -1113,4 +1114,77 @@ private[ml] object RandomForest extends Logging {
     }
   }
 
+  /**
+   * Given a Random Forest model, compute the importance of each feature.
+   * This generalizes the idea of "Gini" importance to other losses,
+   * following the explanation of Gini importance from "Random Forests" documentation
+   * by Leo Breiman and Adele Cutler, and following the implementation from scikit-learn.
+   *
+   * This feature importance is calculated as follows:
+   *  - Average over trees:
+   *     - importance(feature j) = sum (over nodes which split on feature j) of the gain,
+   *       where gain is scaled by the number of instances passing through node
+   *     - Normalize importances for tree based on total number of training instances used
+   *       to build tree.
+   *  - Normalize feature importance vector to sum to 1.
+   *
+   * Note: This should not be used with Gradient-Boosted Trees.  It only makes sense for
+   *       independently trained trees.
+   * Note: This is returned as a Map since models do not store the number of features.
+   *       That should be corrected in the future.
+   * @param trees  Unweighted forest of trees
+   * @return  Feature importance values.  Returned as map from feature index to importance.
+   */
+  private[ml] def featureImportances(trees: Array[DecisionTreeModel]): Map[Int, Double] = {
+    val totalImportances = new OpenHashMap[Int, Double]()
+    trees.foreach { tree =>
+      // Aggregate feature importance vector for this tree
+      val importances = new OpenHashMap[Int, Double]()
+      computeFeatureImportance(tree.rootNode, importances)
+      // Normalize importance vector for this tree, and add it to total.
+      val treeNorm = tree.rootNode.impurityStats.count
```

Comment:
I see. They way I look at it is that (a) is a must, the impurity at each node is weighted by #instances divided total#instances, otherwise the importance score would be too big. Then, for normalization, there are several ways: (1) by making the sum equal to 1 (like ones provided in sklearn or in some R packages), or (2) by making the highest importance scaled to 1 (like ones provided in Hastie's book). Anyway, the code looks good to me :).